Remove all backup files and folders before uploading a package version

ellensi opened the pull request:
### Purpose

References:
- [MAGN-7676](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7676) Do not include backup folders when uploading a package
- #4699 Disregard backup folders when uploading a new package version

When uploading a package, PackageManager copies selected files from their respective original locations to the target package folder and then zips the whole package folder for upload.

For the case when the user performs Publish Version, the original location is the same as the target package folder. PackageManager automatically selects and sorts the files in the old folder (skipping the backup files), but the files that have been skipped during the listing here will still reside in this folder, hence they will be included for zip and upload.

An implementation using `IFileSystem` has been added to delete the files in the package folder which are already not selected. All sub-folders named "backup" will also be deleted entirely before zipping the package folder.

A test case for this defect has also been added.

### Declarations

- [x] The level of testing this PR includes is appropriate
- [x] All tests pass using the self-service CI.

### Reviewers

@pboyer @Benglin 

### FYIs

@riteshchandawar @kronz

pboyer commented on src/DynamoPackages/PackageDirectoryBuilder.cs:
```diff
@@ -92,6 +93,18 @@ private void RemoveDyfFiles(IEnumerable<string> filePaths, IDirectoryInfo dyfDir
+
+            // Remove all backup folders
+            Directory.GetDirectories(dir.FullName, Configuration.Configurations.BackupFolderName, SearchOption.AllDirectories)
+                .ToList().ForEach(x => Directory.Delete(x));
```
`ToList().ForEach(...`

is an antipattern. It allocates an unnecessary list. Just use `foreach`. :)

pboyer commented on src/DynamoPackages/PackageDirectoryBuilder.cs:
```diff
@@ -92,6 +93,18 @@ private void RemoveDyfFiles(IEnumerable<string> filePaths, IDirectoryInfo dyfDir
+        {
+            // Remove all files which are not listed in the files list
+            Directory.GetFiles(dir.FullName, "*", SearchOption.AllDirectories)
+                .Where(x => !x.EndsWith("pkg.json") && !filePaths.Contains(x)).ToList()
```
Take a look at the `IFileSystem` type. Please use it for all file system operations. You may need to extend it. It makes it easy to write tests for packages that do not mutate the file system.

pboyer commented:
Please add a unit test in the style of the existing unit tests. There are quite a few.

ellensi commented:
I understand your point, @pboyer, thanks a lot for your feedback!

I'm afraid the `IFileSystem` does not deal with enumerating real files in the directory, as we need to enumerate and remove unselected files and backup folders before zipping the package folder.

A different solution is to modify the way PackageDirectoryBuilder works, by creating a temporary folder which contains all files to be zipped. This might be a big change anyway.

Please let me hear your voice about this, and thanks a lot!

Benglin commented on src/DynamoPackages/Interfaces/IFileSystem.cs:
```diff
@@ -7,8 +8,12 @@ namespace Dynamo.PackageManager
     public interface IFileSystem
     {
+        IEnumerable<string> EnumerateAllFiles(string dir);
+        IEnumerable<string> EnumerateAllDirectories(string dir);
```
Should these be aligned better with the underlying system APIs and be called `GetFiles` and `GetDirectories`?

Benglin commented on test/Libraries/PackageManagerTests/PackageDirectoryBuilderTests.cs:
```diff
@@ -185,6 +185,35 @@ public void BuildPackageDirectory_DeletesTheOriginalFiles()
+                "C:/pkg/extra/file3.txt", "C:/pkg/extra/backup/file3.txt.backup", "C:/pkg/extra/subfolder/file4.dwg" });
+            fs.SetAllDirectories(new List<string>() {
+                "C:/pkg/bin", "C:/pkg/dyf", "C:/pkg/dyf/backup", "C:/pkg/extra",
+                "C:/pkg/extra/backup", "C:/pkg/extra/subfolder" });
```
Can we rename one of these to `Backup` instead of `backup`?

Benglin commented on src/DynamoPackages/Package.cs:
```diff
@@ -205,7 +205,9 @@ public void EnumerateAdditionalFiles()
-                .Where(x => !x.ToLower().EndsWith(".dyf") && !x.ToLower().EndsWith(".dll") && !x.ToLower().EndsWith("pkg.json") && !x.ToLower().EndsWith(".backup"))
+                .Where(x => !x.ToLower().EndsWith(".dyf") && !x.ToLower().EndsWith(".dll") &&
+                       !x.ToLower().EndsWith("pkg.json") && !x.ToLower().EndsWith(".backup") &&
+                       !x.ToLower().Contains(@"\" + Configuration.Configurations.BackupFolderName + @"\"))
```
I tend to think I prefer this (more readable), but that's entirely up to you:

```csharp
// Do this only once so that's easier to maintain or update.
var backupFolderName = @"\" + Configuration.Configurations.BackupFolderName + @"\";


var nonDyfDllFiles = Directory.EnumerateFiles(...)
.Where((x) =>
{
    var path = x.ToLower();

    return = !path.EndsWith(".dyf") &&
        !path.EndsWith(".dll") &&
        !path.EndsWith("pkg.json") &&
        !path.EndsWith(".backup") &&
        !path.Contains(backupFolderName);
})
```

Benglin commented:
Thanks for helping with this, @ellensi! Few things and it'll be over to @pboyer :)

ellensi commented:
Hi @pboyer,

Addressing your previous comments, I have introduced `GetFiles()` and `GetDirectories()` methods under `IFileSystem` and added a unit test case for this defect.

Please let me know if you have any more comments about this. Thank you!

pboyer commented:
Wonderful! Thanks @ellensi! Sorry to be a stickler about this.
